### ISSUE TITLE
Include a custom resolver in Jest preset

### ISSUE THREAD

**Issue by SimenB** [react-native-community/discussions-and-proposals#512]
### Introduction

Metro resolves a custom field in `package.json` called `react-native`. This is not supported by Jest's default resolver - just like e.g. `module` isn't. This means different modules are resolved when running tests than when `metro` is bundling code.

### Details

The Jest preset in `react-native` could ship a [`resolver`](https://jestjs.io/docs/configuration#resolver-string) which uses e.g. `packageFilter` to prioritize `react-native` over `main` etc..

While the long term solution to this issue is probably `exports` and passing a `react-native` condition, Metro ([currently](https://github.com/facebook/metro/issues/670#issuecomment-1226005278)) doesn't support `exports` and regardless, that'll leave all existent modules in the dust.

Note that this field seems to be undocumented? Sorta mentioned [here](https://facebook.github.io/metro/docs/resolution#getpackagemainpath-string--string), but not really 😅 

Like with #509, I'd be happy to help review and/or write the code if this is accepted.

### Discussion points

Different resolution of module leads to more different code than necessary running in tests vs production code, making them more brittle and less useful. Providing a custom `resolver` in the preset could help alleviate this.

**Comment by thymikee** [comment]
Metro team has plans to support `exports`. cc @motiz88 @huntie

**Comment by SimenB** [comment]
Yeah, I linked to the comment where the team stated it was on the roadmap. Regardless, that'd still leave 100% of existing modules behind, so I think support for `react-native` should be added

**Comment by NickGerleman** [comment]
Jest's default resolver is also missing some of the smarts for out-of-tree platforms (so the wrong platform's implementation could be included).  I remember adding a custom resolver in the RNW repo for this a while back, but IIRC @tido64 did something better as part of the rnx-kit efforts which should be publicly usable.

**Comment by kelset** [comment]
Maybe you are thinking of this? https://github.com/microsoft/rnx-kit/tree/main/packages/typescript-react-native-resolver

**Comment by tido64** [comment]
I think @NickGerleman is referring to https://github.com/microsoft/rnx-kit/tree/main/packages/jest-preset. It contains logic for picking out the correct platform if within an out-of-tree platform, and provides affordances for specifying platforms otherwise.
